src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -199,34 +199,41 @@ private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExCon
         {
             GetAddrInfoExState state = GetAddrInfoExState.FromHandleAndFree(context->QueryStateHandle);
 
+            object result;
             try
             {
                 CancellationToken cancellationToken = state.UnregisterAndGetCancellationToken();
 
                 if (errorCode == SocketError.Success)
                 {
                     IPAddress[] addresses = ParseAddressInfoEx(context->Result, state.JustAddresses, out string? hostName);
-                    state.SetResult(state.JustAddresses ? (object)
+                    result = state.JustAddresses ?
                         addresses :
                         new IPHostEntry
                         {
                             HostName = hostName ?? state.HostName,
                             Aliases = Array.Empty<string>(),
                             AddressList = addresses
-                        });
+                        };
                 }
                 else
                 {
                     Exception ex = (errorCode == (SocketError)Interop.Winsock.WSA_E_CANCELLED && cancellationToken.IsCancellationRequested)
-                        ? (Exception)new OperationCanceledException(cancellationToken)
+                        ? new OperationCanceledException(cancellationToken)
                         : new SocketException((int)errorCode);
-                    state.SetResult(ExceptionDispatchInfo.SetCurrentStackTrace(ex));
+                    result = ExceptionDispatchInfo.SetCurrentStackTrace(ex);
                 }
             }
+            catch (Exception ex)
+            {
+                result = ex;
+            }
             finally
             {
                 state.Dispose();
             }
+
+            state.SetResult(result);
```

Comment:
It'd be worth a comment both in Dispose and here about why it's valid to be using state after it's been Dispose'd (with the comment in Dispose making it clear that a future edit shouldn't disrupt this being ok).

Reply:
Added a comment.